### PR TITLE
Closes #3267 Remove Cache Lifespan “Minutes” Frequency

### LINKED ISSUE #3267: Remove Cache Lifespan “Minutes” Frequency
## What are we proposing to do? 

With this minor enhancement, we will remove the “minutes” option as a frequency from the Cache Lifespan choices

## Why are we doing this? 

Clearing the cache every few minutes can mean that WP Rocket is constantly generating cache files. This can be bad for performance and can cause high CPU issues. 

<p></p>

**Note**: If customers have these use cases, where they need a really short cache, the better approach would be to [<u>disable page cache generation</u>](https://docs.wp-rocket.me/article/61-disable-page-caching) and rely on a server cache like Varnish.   

WP Rocket will soon be interacting with our upcoming SASS API’s on a page by page basis. Websites using the API every few minutes could overload our API and cost us.   

Furthermore, as per MixPanel data, this option is only used by about 1.34% of our customers. 

[<u>https://mixpanel.com/s/1lAQmM</u>](https://mixpanel.com/s/1lAQmM)

<p></p>

If we break down that 1.34% further down, a sizable number of users use values greater than 60 minutes. These use cases can be covered by our “hours” and “days” frequency. 

[<u>https://mixpanel.com/s/2QNh1J</u>](https://mixpanel.com/s/2QNh1J)

## User expectations and experience. How will we do this feature?

This update will be transparent for over 98.5% of the users. They will not be affected in any way.   

For the remaining users, we will port them to the closest frequency as per the following proposal:   

- Any user set to less than 60 minutes can be set to 1 hour

- Users using values greater than 60 minutes can be ported to the closest hour. Example: Values from 60-119 can be ported to 1 hour, values from 120-179 can be ported to 2 hours, and so on. 

## Technical Specs

Please refer to the user experience. 

## Risks

The primary risk is that the user might not know about this change and they might miss all the announcements we make. This will impact about 1% of our users (users with values less than 60 minutes). We can educate them as much as we can. However, if they insist on having their cache cleared every 10 minutes, we can assist them with [<u>setting up a cron job</u>](https://docs.wp-rocket.me/article/494-how-to-clear-cache-via-cron-job) that clears their cache every 10 minutes. 

## Acceptance Criteria

- The UI element for “minutes” is removed. 

- When a user updates from old version to new with “hours”, or “days” as the option, nothing changes for them. 

- If the user updates with minutes set as the cache lifespan option, they are set to “hours” based on the criteria in the user experience. i.e. <60 minutes is set to 1 hour and >60 minutes is approximated to the lowest hour.

- Cache lifespan can be set to 0 for infinite cache lifespan means that the cache is no longer deleted. 

## Documentation and Translations

- None of our documentation mentions “minutes” as far as I could find. Don’t think we need to update anything. 

- This issue is no longer relevant and can be closed - [<u>https://github.com/wp-media/wp-rocket/issues/2202</u>](https://github.com/wp-media/wp-rocket/issues/2202)

- We might need to inform our users about this change via an email, release post, beacon message, and changelog.

### PR DESCRIPTION
**Pull request by engahmeds3ed**
Closes #3267 

### Code changes

- [x] Move `inc/classes/Cache/class-expired-cache-purge.php` and `inc/classes/subscriber/Cache/class-expired-cache-purge-subscriber.php` to the new architecture. Group them in `inc/Engine/Cache/PurgeExpired/`, rename the files & classes

- [x] Add class alias for the moved classes above, in the `deprecated/3.8.php` file

#### `WP_Rocket\Engine\Cache\PurgeExpired\PurgeExpiredCache`
- [x] Add a method `update_lifespan_value()` to update the lifespan value if it was using the minutes value before
- [x] Add tests

#### `WP_Rocket\Engine\Cache\PurgeExpired\Subscriber`
- [x] Remove `clean_expired_cache_scheduled_event()` method and the associated event
- [x] Add a callback on `wp_rocket_upgrader`, calling `PurgeExpiredCache::update_lifespan_value()`
- [x] Add tests

#### `WP_Rocket\Engine\Admin\Settings\Page`
- [x] Update the `purge_cron_interval` index, to remove the `MINUTE_IN_SECONDS` choice

#### `WP_Rocket\Engine\Admin\Settings\Settings`
- [x] Update `sanitize_callback()`, to remove the `MINUTE_IN_SECONDS` choice in`$allowed_cron_units`
- [x] Remove the conditional checking for a value < 10 minutes

### CODE REVIEW

File: inc/Engine/Cache/PurgeExpired/PurgeExpiredCache.php [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
  * @since  3.4
- * @author Grégory Viguier
  */
-class Expired_Cache_Purge {
+class PurgeExpiredCache {
 	/**
 	 * Path to the global cache folder.
 	 *
 	 * @since  3.4
 	 * @access private
```
Comment (Tabrisrp) [inline inc/Engine/Cache/PurgeExpired/PurgeExpiredCache.php:16]: You can remove all the `@access` in this class too

File: inc/Engine/Cache/PurgeExpired/Subscriber.php [hunk 2]
```diff
@@ -28,7 +25,6 @@ class Expired_Cache_Purge_Subscriber implements Subscriber_Interface {
 	 *
 	 * @since  3.4
 	 * @access private
```
Comment (Tabrisrp) [inline inc/Engine/Cache/PurgeExpired/Subscriber.php:27]: Same here, remove all `@access`

File: inc/Engine/Cache/PurgeExpired/PurgeExpiredCache.php [hunk 3]
```diff
@@ -390,4 +379,31 @@ private function is_dir_empty( $dir_path ) {
 
 		return true;
 	}
+
+	/**
+	 * Update lifespan option to convert old minutes to hours.
+	 *
+	 * @since 3.8
+	 *
+	 * @param int $old_lifespan Old value in minutes.
```
Comment (Tabrisrp) [inline inc/Engine/Cache/PurgeExpired/PurgeExpiredCache.php:388]: missing the second param doc

File: inc/Engine/Cache/ServiceProvider.php [hunk 4]
```diff
@@ -48,5 +50,12 @@ public function register() {
 		$this->getContainer()->share( 'admin_cache_subscriber', 'WP_Rocket\Engine\Cache\AdminSubscriber' )
 			->withArgument( $this->getContainer()->get( 'advanced_cache' ) )
 			->withArgument( $this->getContainer()->get( 'wp_cache' ) );
+
+		$options = $this->getContainer()->get( 'options' );
+		$this->getContainer()->add( 'expired_cache_purge', 'WP_Rocket\Engine\Cache\PurgeExpired\PurgeExpiredCache' )
+			 ->withArgument( rocket_get_constant( 'WP_ROCKET_CACHE_PATH' ) );
+		$this->getContainer()->share( 'expired_cache_purge_subscriber', 'WP_Rocket\Engine\Cache\PurgeExpired\Subscriber' )
+			 ->withArgument( $options )
+			 ->withArgument( $this->getContainer()->get( 'expired_cache_purge' ) );
```
Comment (Tabrisrp) [inline inc/Engine/Cache/ServiceProvider.php:59]:
```suggestion
		$this->getContainer()->add( 'expired_cache_purge', 'WP_Rocket\Engine\Cache\PurgeExpired\PurgeExpiredCache' )
			 ->withArgument( rocket_get_constant( 'WP_ROCKET_CACHE_PATH' ) );
		$this->getContainer()->share( 'expired_cache_purge_subscriber', 'WP_Rocket\Engine\Cache\PurgeExpired\Subscriber' )
			 ->withArgument( $this->getContainer()->get( 'options' ) )
			 ->withArgument( $this->getContainer()->get( 'expired_cache_purge' ) );
```

No need to assign it to a variable if it's used only once

**Review by Tabrisrp — CHANGES_REQUESTED** [review]

**Review by Tabrisrp — APPROVED** [review]

**Comment by Tabrisrp** [comment]
@wp-media/qa Ready for QA

**Comment by piotrbak** [comment]
@Tabrisrp @engahmeds3ed I'm just waiting to confirm that the expired purge works as expected (need to wait one hour now :D ).

When downgrading to the previous version, we're receiving this PHP notice in the WP Debug log:
`[16-Nov-2020 17:59:12 UTC] PHP Notice:  Undefined property: stdClass::$class in C:\Users\pbak\Local Sites\version\app\public\wp-content\plugins\wp-rocket\views\settings\page-sections\dashboard.php on line 76`

What do you think?

https://snippi.com/s/cpye6nc

**Review by piotrbak — APPROVED** [review]
@Tabrisrp As we talked, if this notice is not related to this change, then other tests passed.

**Pull request merged**